### PR TITLE
Rewrite detect_line_breaks_inner

### LINKED ISSUE #346: Rewrite `detect_line_breaks_inner`
**Is your feature request related to a problem? Please describe.**
This function has become a mess. Rewrite it, preferably in a functional style without mutable variables.

**Describe the solution you'd like**
A reimplementation that is easy to read.

### PR DESCRIPTION
**Pull request by Xophmeister**
Resolves #346

Note: The flattening of the syntax tree can be factored out, to avoid having to do it twice (for single line breaks and blank lines). I will do this refactor in a separate PR.

### CODE REVIEW

File: topiary/src/atom_collection.rs [hunk 1]
```diff
@@ -773,63 +773,61 @@ fn detect_multi_line_nodes(node: &Node) -> HashSet<usize> {
...
-                node.id()
-            );
+) -> (HashSet<usize>, HashSet<usize>) {
+    // Flatten the tree, depth-first, into a vector of nodes
+    let mut walker = node.walk();
+    let mut dfs_nodes: Vec<Node> = Vec::new();
+
+    // NOTE Could this be written in functional style?
+    // Either way, it can be factored out to avoid running twice.
+    'walk: loop {
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:793]: It looks cleaner to do this recursively, like we do several places, e.g. `collect_leafs_inner`.

File: topiary/src/atom_collection.rs [hunk 2]
```diff
@@ -773,63 +773,61 @@ fn detect_multi_line_nodes(node: &Node) -> HashSet<usize> {
...
+                log::debug!(
+                    "There are at least {} line breaks between {:?} and {:?}",
+                    minimum_line_breaks,
+                    alpha.id(),
+                    omega.id()
+                );
+
+                // We only need the node IDs and, to preserve the API,
+                // flip the output to ("before", "after") sets.
+                return Some((omega.id(), alpha.id()));
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:825]: Maybe even rename alpha and omega to before and after?

File: topiary/src/atom_collection.rs [hunk 3]
```diff
@@ -773,63 +773,61 @@ fn detect_multi_line_nodes(node: &Node) -> HashSet<usize> {
...
+            if next >= last + minimum_line_breaks {
+                log::debug!(
+                    "There are at least {} line breaks between {:?} and {:?}",
+                    minimum_line_breaks,
+                    alpha.id(),
+                    omega.id()
+                );
+
+                // We only need the node IDs and, to preserve the API,
+                // flip the output to ("before", "after") sets.
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:824]: Shouldn't this say "to (after, before)"?

File: topiary/src/atom_collection.rs [hunk 4]
```diff
@@ -773,63 +773,61 @@ fn detect_multi_line_nodes(node: &Node) -> HashSet<usize> {
...
-        previous_end = end;
-        nodes_with_breaks_before.extend(before);
-        nodes_with_breaks_after.extend(after);
-    }
+    // Zip the flattened vector with its own tail => Iterator of pairs of adjacent nodes
+    // Filter this by the threshold distance between pair components
+    // Unzip into "after" and "before" sets, respectively
+    dfs_nodes
+        .iter()
+        .zip(dfs_nodes[1..].iter())
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:810]: Neat!

File: topiary/src/atom_collection.rs [hunk 5]
```diff
@@ -773,63 +773,61 @@ fn detect_multi_line_nodes(node: &Node) -> HashSet<usize> {
 }
 
 fn detect_blank_lines_before(node: &Node) -> HashSet<usize> {
-    detect_line_breaks_inner(node, 2, None, None).0
+    detect_line_breaks_inner(node, 2).0
```
Comment (torhovland) [inline topiary/src/atom_collection.rs:776]: This will be even nicer if the output from `detect_line_breaks_inner` is a struct rather than a tuple.

**Review by torhovland — APPROVED** [review]

File: topiary/src/atom_collection.rs (hunk 3 above)
Comment (Xophmeister) [reply topiary/src/atom_collection.rs:824]: It should...I confused myself :sweat_smile: 

In my subsequent PR, I've just changed the API, so there's no confusion. Using a struct will make it even more explicit :+1:

File: topiary/src/atom_collection.rs (hunk 1 above)
Comment (Xophmeister) [reply topiary/src/atom_collection.rs:793]: I am factoring this out in a subsequent PR. I'll look into making it recursive there...

**Pull request merged**